src/adapter/src/coord/timestamp_selection.rs

```diff
@@ -208,6 +208,45 @@ pub trait TimestampProvider {
 
     async fn oracle_read_ts(&self, timeline: &Timeline) -> Option<Timestamp>;
 
+    fn get_timeline(timeline_context: &TimelineContext) -> Option<Timeline> {
+        let timeline = match timeline_context {
+            TimelineContext::TimelineDependent(timeline) => Some(timeline.clone()),
+            // We default to the `Timeline::EpochMilliseconds` timeline if one doesn't exist.
+            TimelineContext::TimestampDependent => Some(Timeline::EpochMilliseconds),
+            TimelineContext::TimestampIndependent => None,
+        };
+
+        timeline
+    }
+
+    /// Returns a `Timeline` whose timestamp oracle we have to use to get a
+    /// linearized read timestamp, _iff_ linearization is needed.
+    fn get_linearized_timeline(
+        isolation_level: &IsolationLevel,
+        when: &QueryWhen,
+        timeline_context: &TimelineContext,
+    ) -> Option<Timeline> {
+        let timeline = Self::get_timeline(timeline_context);
+
+        // In order to use a timestamp oracle, we must be in the context of some timeline. In that
+        // context we would use the timestamp oracle in the following scenarios:
+        // - The isolation level is Strict Serializable and the `when` allows us to use the
+        //   the timestamp oracle (ex: queries with no AS OF).
+        // - The `when` requires us to use the timestamp oracle (ex: read-then-write queries).
+        let linearized_timeline = match &timeline {
+            Some(timeline)
+                if when.must_advance_to_timeline_ts()
+                    || (when.can_advance_to_timeline_ts()
+                        && isolation_level == &IsolationLevel::StrictSerializable) =>
+            {
+                Some(timeline.clone())
+            }
+            _ => None,
+        };
+
+        linearized_timeline
+    }
```

Comment:
There's probably nothing to do about that, but I thought I'd call it out in case you had an idea.

Reply:
It's related to https://github.com/MaterializeInc/materialize/pull/23421#discussion_r1404502212: `TimestampProvider`, which `Coordinator` implements is `#[async_trait(?Send)]`, because `Coordinator` cannot be `Send` and/or `Sync`.

Also very annoying, but what I did is essentially pull out the thing that can be done asynchronously (the `read_ts` call on a shareable oracle). We might get back to something better again further down the road.